ethcore/src/evm/wasm/mod.rs

```diff
@@ -0,0 +1,159 @@
+		use parity_wasm::elements::Deserialize;
+
+		let code = params.code.expect("exec is only called on contract with code; qed");
+
+		trace!(target: "wasm", "Started wasm interpreter with code.len={:?}", code.len());
+
+		let env_instance = self.program.module("env")
+			// prefer explicit panic here
+			.expect("Wasm program to contain env module");
+
+		let env_memory = env_instance.memory(interpreter::ItemIndex::Internal(0))
+			// prefer explicit panic here
+			.expect("Linear memory to exist in wasm runtime");
+
+		if params.gas > ::std::u64::MAX.into() {
+			return Err(evm::Error::Wasm("Wasm interpreter cannot run contracts with gas >= 2^64".to_owned()));
+		}
+		
+		let mut runtime = Runtime::with_params(
+			ext,
+			env_memory,
+			DEFAULT_STACK_SPACE,
+			params.gas.low_u64(),
+		);
+
+		let mut cursor = ::std::io::Cursor::new(&*code);
+
+		let contract_module = wasm_utils::inject_gas_counter(
+			elements::Module::deserialize(
+				&mut cursor
+			).map_err(|err| {
+				evm::Error::Wasm(format!("Error deserializing contract code ({:?})", err))
+			})?
+		);
+
+		let d_ptr = runtime.write_descriptor(
+			call_args::CallArgs::new(
+				params.address,
+				params.sender,
+				params.origin,
+				params.value.value(),
+				params.data.unwrap_or(Vec::with_capacity(0)),
+			)
+		)?;
+
+		{
+			let execution_params = interpreter::ExecutionParams::with_external(
+				"env".into(), 
+				Arc::new(
+					interpreter::env_native_module(env_instance, native_bindings(&mut runtime))
+						.map_err(|err| {
+							// todo: prefer explicit panic here also?
+							evm::Error::Wasm(format!("Error instantiating native bindings: {:?}", err))
+						})?
+				)
+			).add_argument(interpreter::RuntimeValue::I32(d_ptr.as_raw() as i32));
+		
+			let module_instance = self.program.add_module("contract", contract_module, Some(&execution_params.externals))
+				.map_err(|err| {
+					trace!(target: "wasm", "Error adding contract module: {:?}", err);
+					evm::Error::from(RuntimeError::Interpreter(err))
+				})?;
+
+			module_instance.execute_export("_call", execution_params)
+				.map_err(|err| {
+					trace!(target: "wasm", "Error executing contract: {:?}", err);
+					evm::Error::from(RuntimeError::Interpreter(err))
+				})?;
+		}
+
+		let result = result::WasmResult::new(d_ptr);
+		if result.peek_empty(&*runtime.memory())? {
+			trace!(target: "wasm", "Contract execution result is empty.");
+			Ok(GasLeft::Known(runtime.gas_left()?.into()))
+		} else {
+			self.result.clear();
+			// todo: use memory views to avoid copy
+			self.result.extend(result.pop(&*runtime.memory())?);
+			let len = self.result.len();
+			Ok(GasLeft::NeedsReturn { 
+				gas_left: runtime.gas_left()?.into(), 
+				data: ReturnData::new(
+					::std::mem::replace(&mut self.result, Vec::with_capacity(DEFAULT_RESULT_BUFFER)), 
+					0, 
+					len,
+				),
+				apply_state: true,
+			})
+		}
+	}
+}
+
+fn native_bindings<'a>(runtime: &'a mut Runtime) -> interpreter::UserFunctions<'a> {
```

Comment:
It's not because `UserFunctions` has a lifetime parameter, it's because `Runtime` has a lifetime parameter. `&mut Runtime` has two elided lifetime parameters, so when you had the lifetime parameter to the function it elided one of them and not the other. You do need the parameters in this case, sorry. I didn't know `Runtime` took a lifetime parameter.

Basically, it desugars to:

```rust
fn native_bindings<'a, 'b>(runtime: &'a mut Runtime<'b>) -> interpreter::UserFunctions<'???>
```

Since there are two parameters to choose from, it doesn't know which to pass to `UserFunctions`.